enclave/attest_plugin.c

```diff
@@ -0,0 +1,294 @@
+
+    _dump_attestation_plugin_list();
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_unregister_attestation_plugin(
+    oe_quote_customization_plugin_context_t* plugin)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    struct plugin_list_node_t* prev = NULL;
+    struct plugin_list_node_t* cur = g_plugins;
+
+    OE_TRACE_INFO("Calling oe_unregister_attestation_plugin 1");
+
+    // Find the guid and remove it.
+    cur = _find_plugin(&plugin->evidence_format_uuid, &prev);
+    if (cur == NULL)
+    {
+        OE_TRACE_ERROR(
+            "Calling oe_unregister_attestation_plugin failed: "
+            "evidence_format_uuid[%s] was not registered before",
+            plugin->evidence_format_uuid);
+        OE_RAISE(OE_NOT_FOUND);
+    }
+
+    oe_mutex_lock(&g_plugin_list_mutex);
+    if (prev != NULL)
+        prev->next = cur->next;
+    else
+        g_plugins = NULL;
+    oe_mutex_unlock(&g_plugin_list_mutex);
+
+    // Run the unregister hook for the plugin.
+    if (cur->plugin_context->on_unregister(cur->plugin_context) != 0)
+        OE_RAISE(OE_FAILURE);
+
+    _dump_attestation_plugin_list();
+    result = OE_OK;
+
+done:
+    oe_free(cur);
+    return result;
+}
+
+oe_result_t oe_get_attestation_evidence(
+    const uuid_t* evidence_format_uuid,
+    const uint8_t* user_data,
+    size_t user_data_size,
+    uint8_t** evidence_buffer,
+    size_t* evidence_buffer_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    struct plugin_list_node_t* plugin_node = NULL;
+    oe_quote_customization_plugin_context_t* plugin_context = NULL;
+    int ret;
+    uint8_t* plugin_evidence = NULL;
+    size_t plugin_evidence_size = 0;
+    uint8_t* total_evidence_buf = NULL;
+    size_t total_evidence_size = 0;
+    oe_evidence_header_t* header;
+
+    // Find a plugin for attestation type and run its get_evidence.
+    plugin_node = _find_plugin(evidence_format_uuid, NULL);
+    if (plugin_node == NULL)
+        OE_RAISE(OE_NOT_FOUND);
+
+    plugin_context = plugin_node->plugin_context;
+    ret = plugin_context->get_evidence(
+        plugin_context,
+        user_data,
+        user_data_size,
+        &plugin_evidence,
+        &plugin_evidence_size);
+    if (ret != 0)
+    {
+        OE_TRACE_ERROR("get_evidence failed with ret = %d", ret);
+        OE_RAISE(OE_FAILURE);
+    }
+
+    // Wrap the OE header around it.
+    OE_CHECK(oe_safe_add_sizet(
+        sizeof(oe_evidence_header_t),
+        plugin_evidence_size,
+        &total_evidence_size));
+
+    total_evidence_buf = (uint8_t*)oe_malloc(total_evidence_size);
+    if (total_evidence_buf == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    // TODO(akagup): Fix oe report header to only have 1 size parameter.
+    header = (oe_evidence_header_t*)total_evidence_buf;
+    header->version = OE_REPORT_HEADER_VERSION;
+    header->evidence_format_uuid = *evidence_format_uuid;
+    header->tee_evidence_type = OE_TEE_TYPE_CUSTOM;
+    header->tee_evidence_size = 0;
```

Comment:
all the plug-in will have CUSTOM as evidence type?
Is there anyway for a plugin to support SGX or Cyres type?

Reply:
I made them all custom because I wasn't sure what a plugin w/ a SGX/Cyres type meant. Is that just a plugin that supports SGX only?

Reply:
Per our earlier discussion, let's keep it as custom as it is.